tika/server.go

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2017 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tika
+
+import (
+	"context"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"os/exec"
+	"time"
+
+	"golang.org/x/net/context/ctxhttp"
+)
+
+// Server represents a Tika server. Create a new Server with NewServer,
+// start it with Start, and shut it down with the close function returned
+// from Start.
+// There is no need to create a Server for an already running Tika Server
+// since you can pass its URL directly to a Client.
+type Server struct {
+	jar            string
+	url            string // url is derived from port and hostname.
+	port           string
+	hostname       string
+	cancel         func()
+	startupTimeout time.Duration
+}
+
+// URL returns the URL of this Server.
+func (s *Server) URL() string {
+	return s.url
+}
+
+// An Option can be passed to NewServer to configure the Server.
+type Option func(*Server)
+
+// WithHostname returns an Option to set the host of the Server (default localhost).
+func WithHostname(h string) Option {
```

Comment:
I think using options in here is a bit overkill.
Your server always connect somewhere, isn't it? Why would Hostname and Port be optional?

See my comments on `NewServer`.